src/devices_models/devices/common/range_constraint.jl

```diff
@@ -152,6 +152,159 @@ function add_semicontinuous_range_constraints!(
         # TODO: deal with additional expressions terms for services
         # add_services_constraint!(expression_ub, model)
         # add_services_constraint!(expression_lb, model)
+        if JuMP.has_lower_bound(jump_variable[ci_name, t])
+            JuMP.set_lower_bound(jump_variable[ci_name, t], 0.0)
+        end
+        expression_ub = JuMP.AffExpr(0.0, jump_variable[ci_name, t] => 1.0)
+        expression_lb = JuMP.AffExpr(0.0, jump_variable[ci_name, t] => 1.0)
+        limits = get_min_max_limits(device, T, W) # depends on constraint type and formulation type
+        con_ub[ci_name, t] = JuMP.@constraint(
+            container.JuMPmodel,
+            expression_ub <= limits.max * varbin[ci_name, t]
+        )
+        con_lb[ci_name, t] = JuMP.@constraint(
+            container.JuMPmodel,
+            expression_lb >= limits.min * varbin[ci_name, t]
+        )
+    end
+end
+
+@doc raw"""
+Constructs min/max range constraint from device variable and reservation decision variable.
+
+# Constraints
+
+``` varcts[name, t] <= limits.max * (1 - varbin[name, t]) ```
+
+``` varcts[name, t] >= limits.min * (1 - varbin[name, t]) ```
+
+where limits in constraint_infos.
+
+# LaTeX
+
+`` 0 \leq x^{cts} \leq limits^{max} (1 - x^{bin}), \text{ for } limits^{min} = 0 ``
+
+`` limits^{min} (1 - x^{bin}) \leq x^{cts} \leq limits^{max} (1 - x^{bin}), \text{ otherwise } ``
+"""
+function add_reserve_range_constraints!(
+    container::OptimizationContainer,
+    T::Type{InputActivePowerVariableLimitsConstraint},
+    U::Type{<:VariableType},
+    devices::IS.FlattenIteratorWrapper{V},
+    model::DeviceModel{V, W},
+    X::Type{<:PM.AbstractPowerModel},
+    feedforward::Union{Nothing, AbstractAffectFeedForward},
+) where {V <: PSY.Component, W <: AbstractDeviceFormulation}
+    use_parameters = built_for_simulation(container)
+    constraint = T()
+    variable = U()
+    component_type = V
+    time_steps = get_time_steps(container)
+    jump_variable = get_variable(container, variable, component_type)
+    names = [PSY.get_name(d) for d in devices]
+    binary_variables = [ReservationVariable()]
```

Comment:
@kdheepak can we just use the variable. I don't see a reason to have a vector or size 1 here.

Reply:
bump @kdheepak about this